File: logic/src/main/java/com/tp/holdem/logic/HandPredicates.java
package com.tp.holdem.logic;

import com.tp.holdem.logic.model.Card;
import com.tp.holdem.model.common.Honour;
import io.vavr.collection.List;

import java.util.function.Predicate;
import java.util.stream.IntStream;

import static com.tp.holdem.model.common.Honour.*;
import static com.tp.holdem.model.common.Honour.ACE;

class HandPredicates {
	static final Predicate<List<Card>> REGULAR_STRAIGHT = values -> {
		final List<Honour> mapped = values.map(Card::getHonour);
		return IntStream.range(0, mapped.size() - 1)
				.allMatch(index -> mapped.get(index + 1).isRightAfter(mapped.get(index)));
	};
	static final Predicate<List<Card>> ACE_STRAIGHT = values -> values.map(Card::getHonour).containsAll(List.of(ACE, TWO, THREE, FOUR, FIVE));
	static final Predicate<List<Card>> STRAIGHT = ACE_STRAIGHT.or(REGULAR_STRAIGHT);
	static final Predicate<List<Card>> FLUSH = HandFinder::isFlush;
	static final Predicate<List<Card>> STRAIGHT_FLUSH = STRAIGHT.and(FLUSH);
	static final Predicate<List<Card>> ROYAL_FLUSH = REGULAR_STRAIGHT.and(FLUSH).and(cards -> cards.last().getHonour() == ACE);
}

File: logic/src/main/kotlin/com/tp/holdem/logic/hands/HandPredicates.kt
package com.tp.holdem.logic.hands

import com.tp.holdem.model.Card
import io.vavr.collection.List

import java.util.function.Predicate
import java.util.stream.IntStream

import com.tp.holdem.common.model.Honour.*
import com.tp.holdem.common.model.Honour.ACE

internal object HandPredicates {
    val REGULAR_STRAIGHT = Predicate { values: List<Card> ->
        val mapped = values.map { it.honour }
        IntStream.range(0, mapped.size() - 1)
                .allMatch { index -> mapped.get(index + 1).isRightAfter(mapped.get(index)) }
    }
    val ACE_STRAIGHT = Predicate { values: List<Card> -> values.map { it.honour }.containsAll(List.of(ACE, TWO, THREE, FOUR, FIVE)) }
    val STRAIGHT = ACE_STRAIGHT.or(REGULAR_STRAIGHT)
    val FLUSH = Predicate { values: List<Card> -> HandFinder.isFlush(values) }
    val STRAIGHT_FLUSH = STRAIGHT.and(FLUSH)
    val ROYAL_FLUSH = REGULAR_STRAIGHT.and(FLUSH).and { cards -> cards.last().honour == ACE }
}
